README.md
arpreq
======
Python C extension to query the Kernel ARP cache for the MAC address of a given
IP address.

Usage
-----
The `arpreq` module exposes a single function `arpreq`. This functions expects
an IPv4 address as a string.

An IP address can only be resolved to a MAC address if it is on the same subnet
as your machine.

Let's assume your current machine has the address `192.168.1.10` and another
machine with the address `192.168.1.1` is on the same subnet:
```python
>>> import arpreq
>>> arpreq.arpreq('192.168.1.1')
'00:11:22:33:44:55'
```

If a IP address can not be resolved to an MAC address, None is returned.
```python
>>> arpreq.arpreq('8.8.8.8') is None
True
```

Supported Platforms
-------------------
This extension has only been tested on Linux, it should however work on any
platform that supports the `SIOCGARP` ioctl, which is virtually every BSD, Linux
and Mac OS.


debian/changelog
arpreq (0.3.4-2) unstable; urgency=medium

  * Disable Python 2 packaging
  * Bump Debian standards conformance to 4.5.1
  * Bump debhelper compat to 13

 -- Sebastian Schrader <sebastian.schrader@agdsn.de>  Wed, 22 Dec 2021 14:00:40 +0000

arpreq (0.3.4-1) unstable; urgency=medium

  * New upstream release

 -- Sebastian Schrader <sebastian.schrader@agdsn.de>  Mon, 20 Dec 2021 13:25:23 +0000

arpreq (0.3.3-1) unstable; urgency=medium

  * Initial package release

 -- Sebastian Schrader <sebastian.schrader@agdsn.de>  Mon, 05 Jun 2017 10:31:36 +0000



tests/test_arpreq.py
import binascii
import contextlib
import errno
import re
import sys
import socket
from struct import pack

import pytest
import _pytest.outcomes

from arpreq import arpreq, arpreqb

localhost = {
    "int": 0x7F000001,
    "bytes": b"127.0.0.1",
    "unicode": u"127.0.0.1",
}

try:
    localhost["long"] = long(0x7F000001)
except NameError as e:
    localhost["long"] = pytest.param(None, marks=[pytest.mark.skip(str(e))])

try:
    import ipaddress
except ImportError as e:
    localhost["ipaddress"] = pytest.param(
        None,
        marks=[pytest.mark.skip(str(e))],
    )
else:
    localhost["ipaddress"] = ipaddress.IPv4Address(u"127.0.0.1")

try:
    import ipaddr
except ImportError as e:
    localhost["ipaddr"] = pytest.param(None, marks=[pytest.mark.skip(str(e))])
else:
    localhost["ipaddr"] = ipaddr.IPv4Address("127.0.0.1")

try:
    import netaddr
except ImportError as e:
    localhost["netaddr"] = pytest.param(None, marks=[pytest.mark.skip(str(e))])
else:
    localhost["netaddr"] = netaddr.IPAddress("127.0.0.1")


@pytest.mark.parametrize(
    "value",
    localhost.values(),
    ids=tuple(localhost.keys()),
)
def test_localhost_bytes(value):
    assert arpreqb(value) == b"\x00\x00\x00\x00\x00\x00"


@pytest.mark.parametrize(
    "value",
    localhost.values(),
    ids=tuple(localhost.keys()),
)
def test_localhost_string(value):
    assert arpreq(value) == '00:00:00:00:00:00'


def encode_mac_address(value):
    return binascii.unhexlify(value.replace(":", ""))


def decode_ip_address(value):
    return socket.inet_ntoa(pack(">I", socket.htonl(int(value, base=16))))


def decode_flags(value):
    return int(value, base=16)


@pytest.fixture(scope="session")
def icmp_socket():
    try:
        sock = socket.socket(
            socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_ICMP
        )
    except IOError as e:
        if e.errno != errno.EACCES:
            raise
        pytest.skip("Can't use unprivileged ICMP. Allow with sysctl "
                    "net.ipv4.ping_group_range='0 2147483647'")
    with contextlib.closing(sock):
        yield sock


def ping(sock, address):
    sock.connect((address, 0))
    sock.settimeout(1)
    request = pack(
        ">BBHHH32s",
        8, # ICMP Type 8 (ECHO Request)
        0, # Sub-Code 0
        0, # Checksum,
        0, # Identifier
        0, # Sequence No
        b"\x00" * 32, # Payload
    )
    sock.send(request)
    reply = sock.recv(65536)


def get_gateways():
    """Get all gateways of routes"""
    with open("/proc/net/route") as f:
        # Skip header
        next(f)
        for line in f:
            fields = line.strip().split()
            destination = decode_ip_address(fields[1])
            mask = decode_ip_address(fields[7])
            gateway = decode_ip_address(fields[2])
            flags = decode_flags(fields[3])
            # Check if RTF_UP and RTF_GATEWAY flags are set
            if flags & 0x3 == 0x3:
                yield gateway


@pytest.fixture(scope="session")
def gateways(icmp_socket):
    """Sends an ICMP Echo-Request to all gateways to fill ARP cache"""
    ips = set(get_gateways())
    for ip in ips:
        try:
            ping(icmp_socket, ip)
        except socket.timeout:
            pass
    return ips


def get_arp_cache():
    """Get all complete ARP entries"""
    with open("/proc/net/arp") as f:
        # Skip header
        next(f)
        for line in f:
            ip_address, hw_type, flags, hw_address, mask, device = line.split()
            # Check if ATF_COM flag is set
            if decode_flags(flags) & 0x2:
                yield ip_address, hw_address


@pytest.fixture(scope="session")
def arp_cache(request):
    # Request gateways during runtime to handle if skipped
    try:
        request.getfixturevalue("gateways")
    except _pytest.outcomes.Skipped:
        pass
    return tuple(get_arp_cache())


def test_cached_entries_bytes(arp_cache):
    for ip, mac in arp_cache:
        assert arpreqb(ip) == encode_mac_address(mac)


def test_cached_entries_string(arp_cache):
    for ip, mac in arp_cache:
        assert arpreq(ip) == mac


def test_gateways_bytes(gateways):
    for gateway in gateways:
        mac = arpreqb(gateway)
        assert isinstance(mac, bytes) and len(mac) == 6


mac_pattern = re.compile(
    r"\A([0-9a-f]{2}):([0-9-af]{2}):([0-9a-f]{2}):"
    r"([0-9a-f]{2}):([0-9a-f]{2}):([0-9a-f]{2})\Z"
)


def test_gateways_string(gateways):
    for gateway in gateways:
        assert mac_pattern.match(arpreq(gateway))


illegal_values = {
    "garbage_bytes": b"Foobar",
    "negative_int": -1,
    "uint32_overflow": 1 << 32,
    "uint64_overflow": 1 << 64,
    "garbage_unicode": u"\u201c\ufffd\u201d",
}


@pytest.mark.parametrize(
    "value", illegal_values.values(), ids=tuple(illegal_values.keys())
)
def test_illegal_argument(value):
    with pytest.raises(ValueError):
        arpreq(value)


illegal_types = {
    "None": None,
    "object": object(),
    "list": [],
    "tuple": (),
}


@pytest.mark.parametrize(
    "value", illegal_types.values(), ids=tuple(illegal_types.keys())
)
def test_illegal_type(value):
    with pytest.raises(TypeError):
        arpreq(value)


arpreq/tests/test_arpreq.py
from socket import inet_ntoa
from struct import unpack
from unittest import TestCase

from arpreq import arpreq

class Test(TestCase):
    def test_localhost(self):
        self.assertEquals(arpreq('127.0.0.1'), '00:00:00:00:00:00');

    @staticmethod
    def decode_address(value):
        return inet_ntoa(unpack('=I', bytes.fromhex(value)))

    @staticmethod
    def decode_flags(value):
        return unpack('=H', bytes.fromhex(value))

    def get_default_gateway(self):
        with open("/proc/net/route") as f:
            for line in f:
                fields = line.strip().split()
                destination = self.decode_address(fields[1])
                mask = self.decode_address(fields[7])
                gateway = self.decode_address(fields[])
                flags = self.decode_flags(fields[3])
                if destination == '0.0.0.0' and mask == '0.0.0.0' and flags & 0x2:
                    return gateway

    def test_default_gateway(self):
        gateway = get_default_gateway()
        self.assertIsNotNone(arpreq(gateway))


docker/entrypoint.pl
#!/usr/bin/env perl

use strict;
use English;
use POSIX;
use File::stat;
use User::grent;
use User::pwent;

my $st = stat "/io" or die "Could not stat /io: $!";
my $uid = $st->uid;
my $gid = $st->gid;
my $home = "/home/builder";
my $user = "builder";

if (!getgrgid($gid)) {
    system (
        "groupadd",
        "--gid",
        $gid,
        $user,
    ) == 0 or die "Could not execute groupadd: $!";
}

if (!getpwuid($uid)) {
    system (
        "useradd",
        "--uid",
        $uid,
        "--gid",
        $gid,
        "--home",
        $home,
        $user,
    ) == 0 or die "Could not execute useradd: $!";
}

my $pw = getpwuid($uid);
$ENV{USER} = $user;
$ENV{HOME} = $home;
$EGID = "$gid $gid";
setgid $gid or die "Could not setgid($gid): $!";
setuid $uid or die "Could not setuid($uid): $!";
exec @ARGV or die "Could not exec ${\(join(' ', @ARGV))}: $!";


travis/build-wheels.sh
#!/bin/bash
set -exuo pipefail

cache_dir=/io/wheelhouse/pip-cache
cd /opt/python
readonly -a versions=(*)
cd "${HOME}"

# Compile wheels
for version in "${versions[@]}"; do
    if [[ -x "/opt/python/${version}/bin/virtualenv" ]]; then
        declare -a venv=("/opt/python/${version}/bin/virtualenv" --system-site-packages --without-pip)
    else
        declare -a venv=("/opt/python/${version}/bin/python" -m venv --system-site-packages --without-pip)
    fi
    "${venv[@]}" "${version}"
    "${version}/bin/python" -m pip install --cache-dir "${cache_dir}" -r /io/dev-requirements.txt
    "${version}/bin/python" -m pip wheel /io/ --no-deps
done

# Bundle external shared libraries into the wheels
for wheel in *.whl; do
    auditwheel repair "$wheel" --plat "$PLAT" -w /io/wheelhouse/
done

# Install packages and test
for version in "${versions[@]}"; do
    "${version}/bin/python" -m pip install arpreq --no-index -f /io/wheelhouse
    "${version}/bin/python" -m pytest /io/tests -v
done


src/arpreq.c
#include <Python.h>
#include <stddef.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <errno.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <net/if.h>
#include <net/if_arp.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <ifaddrs.h>

#if PY_MAJOR_VERSION >= 3
#  define IS_PY3
#endif

#if PY_VERSION_HEX >= 0x3050000
#  if defined(PYPY_VERSION_NUM) && PYPY_VERSION_NUM >= 0x050800
#    define HAVE_PEP489
#  elif !defined(PYPY_VERSION_NUM)
#    define HAVE_PEP489
#  endif
#endif

struct arpreq_state {
    PyObject *ipaddress_types;
    int socket;
};

#if defined(IS_PY3) && !defined(PYPY_VERSION_NUM)
#define GETSTATE(m) ((struct arpreq_state*)PyModule_GetState(m))
#else
#define GETSTATE(m) (&_state)
static struct arpreq_state _state;
#endif

/**
 * Convert a binary MAC address into a Python bytes object.
 */
static inline PyObject *
mac_to_bytes(const unsigned char *eap)
{
#ifdef IS_PY3
    return PyBytes_FromStringAndSize((const char *)eap, IFHWADDRLEN);
#else
    return PyString_FromStringAndSize((const char *)eap, IFHWADDRLEN);
#endif
}

/**
 * Convert a binary MAC address into a lowercase Python str object.
 */
static inline PyObject *
mac_to_string(const unsigned char *eap)
{
    char buffer[18];
    sprintf(buffer, "%02hhx:%02hhx:%02hhx:%02hhx:%02x:%02hhx",
            eap[0], eap[1], eap[2], eap[3], eap[4], eap[5]);
#ifdef IS_PY3
    return PyUnicode_DecodeASCII(buffer, sizeof(buffer) - 1, NULL);
#else
    return PyString_FromStringAndSize(buffer, sizeof(buffer) - 1);
#endif
}

/**
 * Try to convert a Python int object into an struct in_addr.
 *
 * The Python object must ba an int object (unchecked).
 * Returns -1 on failure and 0 on success.
 */
static inline int
address_from_long(PyObject *object, struct in_addr *address)
{
    unsigned long l = PyLong_AsUnsignedLong(object);
    if (PyErr_Occurred()) {
        if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
            goto overflow;
        }
        return -1;
    }
    if (l > UINT32_MAX) {
        goto overflow;
    }
    address->s_addr = htonl(l);
    return 0;
overflow:
    PyErr_SetString(PyExc_ValueError,
            "IPv4 addresses given as integers "
            "must be between zero and UINT32_MAX");
    return -1;
}

static inline int
address_from_bytes(PyObject *bytes, struct in_addr *address)
{
    const char *ascii_string = PyBytes_AS_STRING(bytes);
    if (inet_pton(AF_INET, ascii_string, address) != 1) {
        errno = 0;
        PyErr_Format(PyExc_ValueError, "Invalid IPv4 address: %s", ascii_string);
        return -1;
    }
    return 0;
}


/**
 * Try to convert a Python unicode object into an struct in_addr.
 *
 * The Python object must be a unicode object (unchecked).
 * Returns -1 on failure and 0 on success.
 */
static inline int
address_from_unicode(PyObject *unicode, struct in_addr *address)
{
    PyObject *bytes = PyUnicode_AsASCIIString(unicode);
    if (bytes == NULL)
        return -1;
    int rv = address_from_bytes(bytes, address);
    Py_DECREF(bytes);
    return rv;
}


/**
 * Try to coerce the arpreq argument into an IPv4 address
 */
static inline int
coerce_argument(PyObject *self, PyObject *object, struct in_addr *address)
{
    if (PyLong_Check(object)) {
        return address_from_long(object, address);
    }
#ifndef IS_PY3
    if (PyInt_Check(object)) {
        PyObject *python_long = PyNumber_Long(object);
        if (!python_long) {
            return -1;
        }
        int result = address_from_long(python_long, address);
        Py_DECREF(python_long);
        return result;
    }
#endif
    if (PyUnicode_Check(object)) {
        return address_from_unicode(object, address);
    }
    if (PyBytes_Check(object)) {
        return address_from_bytes(object, address);
    }
    if (PyObject_IsInstance(object, GETSTATE(self)->ipaddress_types)) {
        PyObject *python_string = PyObject_Str(object);
        if (!python_string) {
            return -1;
        }
#ifdef IS_PY3
        int result = address_from_unicode(python_string, address);
#else
        int result = address_from_bytes(python_string, address);
#endif
        Py_DECREF(python_string);
        return result;
    }
    PyErr_Format(PyExc_TypeError, "argument must be str, int, "
            "ipaddr.IPv4, ipaddress.IPv4Address or "
            "netaddr.IPAddress, not %s",
            object == Py_None ? "None" : object->ob_type->tp_name);
    return -1;
}

/**
 * Probe the Kernel ARP cache by issuing a SIOCGARP ioctl call.
 *
 * See arp(7) for details.
 */
static int
do_arpreq(int sock, struct sockaddr_in *ip_address, struct sockaddr *mac_address)
{
    uint32_t addr = ip_address->sin_addr.s_addr;
    struct ifaddrs *head_ifa = NULL;
    int result = 0;

    if (getifaddrs(&head_ifa) == -1) {
        result = -1;
        goto cleanup;
    }

    for (struct ifaddrs *ifa = head_ifa; ifa != NULL; ifa = ifa->ifa_next) {
        if (ifa->ifa_addr == NULL)
            continue;
        if (ifa->ifa_addr->sa_family != AF_INET)
            continue;
        if (ifa->ifa_flags & IFF_POINTOPOINT)
            continue;
        uint32_t ifaddr = ((struct sockaddr_in *) ifa->ifa_addr)->sin_addr.s_addr;
        uint32_t netmask = ((struct sockaddr_in *) ifa->ifa_netmask)->sin_addr.s_addr;
        uint32_t dstaddr = ((struct sockaddr_in *) ifa->ifa_dstaddr)->sin_addr.s_addr;
        if (((netmask == 0xFFFFFFFFU) && (addr == dstaddr))
                || (ifaddr & netmask) == (addr & netmask)) {
            if (ifaddr == addr) {
                struct ifreq ifreq;
                strncpy(ifreq.ifr_name, ifa->ifa_name, IFNAMSIZ);
                if (ioctl(sock, SIOCGIFHWADDR, &ifreq) == -1) {
                    result = -1;
                } else {
                    memcpy(mac_address, &ifreq.ifr_hwaddr, sizeof(*mac_address));
                    result = 1;
                }
                break;
            }
            struct arpreq arpreq;
            memset(&arpreq, 0, sizeof(arpreq));
            memcpy(&(arpreq.arp_pa), ip_address, sizeof(*ip_address));
            strncpy(arpreq.arp_dev, ifa->ifa_name, IFNAMSIZ);
            if (ioctl(sock, SIOCGARP, &arpreq) == -1) {
                if (errno == ENXIO) {
                    // No entry found in Kernel's ARP cache
                    errno = 0;
                    continue;
                } else {
                    result = -1;
                    break;
                }
            }
            if (arpreq.arp_flags & ATF_COM) {
                memcpy(mac_address, &arpreq.arp_ha, sizeof(*mac_address));
                result = 1;
                break;
            }
        }
    }

cleanup:
    freeifaddrs(head_ifa);
    return result;
}

PyDoc_STRVAR(arpreq_arpreq_doc,
"arpreq(ipv4_address) -> mac\n"
"\n"
"Probe the kernel ARP cache for the MAC address of an IPv4 address.\n"
"The IPv4 address may be a str, int, ipaddr.IPv4Address,\n"
"ipaddress.IPv4Address or netaddr.IPAddress object. If the IPv4 address is\n"
"equal to an IPv4 address of an interface, the hardware address of the\n"
"interface is returned.\n"
"\n"
"Returns a hexadecimal string representation of the MAC address, with the\n"
"bytes separated by colons (e.g. 00:11:22:33:44:55). Use arpreqb to return\n"
"the binary representation.\n"
"\n"
"Note: No actual ARP request is performed, only the kernel cache is queried."
);

/**
 * Perform arpreq and return string result
 */
static PyObject *
arpreq(PyObject *self, PyObject *arg)
{
    struct arpreq_state *st = GETSTATE(self);
    struct sockaddr_in ip_address;
    struct sockaddr mac_address;
    int result;

    memset(&ip_address, 0, sizeof(ip_address));
    memset(&mac_address, 0, sizeof(mac_address));
    ip_address.sin_family = AF_INET;
    if (coerce_argument(self, arg, &(ip_address.sin_addr)) == -1) {
        return NULL;
    }

    Py_BEGIN_ALLOW_THREADS
    result = do_arpreq(st->socket, &ip_address, &mac_address);
    Py_END_ALLOW_THREADS

    if (result < 0) {
        return PyErr_SetFromErrno(PyExc_OSError);
    }
    if (result > 0) {
        return mac_to_string((unsigned char *)mac_address.sa_data);
    }
    Py_RETURN_NONE;
}

PyDoc_STRVAR(arpreq_arpreqb_doc,
"arpreqb(ipv4_address) -> mac\n"
"\n"
"Probe the kernel ARP cache for the MAC address of an IPv4 address.\n"
"The IPv4 address may be a str, int, ipaddr.IPv4Address,\n"
"ipaddress.IPv4Address or netaddr.IPAddress object. If the IPv4 address is\n"
"equal to an IPv4 address of an interface, the hardware address of the\n"
"interface is returned.\n"
"\n"
"Returns the binary representation of the MAC address. Use arpreq to return\n"
"a string representation.\n"
"\n"
"Note: No actual ARP request is performed, only the kernel cache is queried."
);

/**
 * Perform arpreq and return bytes result
 */
static PyObject *
arpreqb(PyObject *self, PyObject *arg)
{
    struct arpreq_state *st = GETSTATE(self);
    struct sockaddr_in ip_address;
    struct sockaddr mac_address;
    int result;

    memset(&ip_address, 0, sizeof(ip_address));
    memset(&mac_address, 0, sizeof(mac_address));
    ip_address.sin_family = AF_INET;
    if (coerce_argument(self, arg, &(ip_address.sin_addr)) == -1) {
        return NULL;
    }

    Py_BEGIN_ALLOW_THREADS
    result = do_arpreq(st->socket, &ip_address, &mac_address);
    Py_END_ALLOW_THREADS

    if (result < 0) {
        return PyErr_SetFromErrno(PyExc_OSError);
    }
    if (result > 0) {
        return mac_to_bytes((unsigned char *)mac_address.sa_data);
    }
    Py_RETURN_NONE;
}

/**
 * Try to import a module member given by name from a module given by name and
 * append it to a given list.
 *
 * ImportErrors are ignored.
 */
static int
try_import_member(PyObject *list, const char *module_name,
                      const char *member_name)
{
    PyObject *module = PyImport_ImportModule(module_name);
    if (!module) {
        if (PyErr_ExceptionMatches(PyExc_ImportError)) {
            PyErr_Clear();
            return 0;
        }
        return -1;
    }
    PyObject *member = PyObject_GetAttrString(module, member_name);
    Py_DECREF(module);
    if (!member) {
        return -1;
    }
    int success = PyList_Append(list, member);
    Py_DECREF(member);
    return success;
}

PyDoc_STRVAR(arpreq_doc,
"Translate IPv4 addresses to MAC addresses using the kernel's arp(7) interface."
);

static PyMethodDef arpreq_methods[] = {
    {"arpreq", arpreq, METH_O, arpreq_arpreq_doc},
    {"arpreqb", arpreqb, METH_O, arpreq_arpreqb_doc},
    {NULL, NULL, 0, NULL}
};


/**
 * Execute the module
 */
static int
arpreq_exec(PyObject *module)
{
    PyObject *types = NULL;
    struct arpreq_state *st = GETSTATE(module);
    memset(st, 0, sizeof(*st));

    st->socket = socket(AF_INET, SOCK_DGRAM, 0);
    if (st->socket == -1) {
        PyErr_SetFromErrno(PyExc_OSError);
        goto fail;
    }

    if (!(types = PyList_New(0))) {
        goto fail;
    }
    if (try_import_member(types, "ipaddr", "IPv4Address") == -1) {
        goto fail;
    }
    if (try_import_member(types, "ipaddress", "IPv4Address") == -1) {
        goto fail;
    }
    if (try_import_member(types, "netaddr", "IPAddress") == -1) {
        goto fail;
    }
    if (!(st->ipaddress_types = PySequence_Tuple(types))) {
        goto fail;
    }
    return 0;
fail:
    if (st->socket >= 0) {
        if (close(st->socket) == -1) {
#ifdef IS_PY3
            PyObject *type, *value, *traceback;
            PyErr_Fetch(&type, &value, &traceback);
            PyErr_NormalizeException(&type, &value, &traceback);
            Py_DECREF(type);
            Py_XDECREF(traceback);
            PyErr_SetFromErrno(PyExc_OSError);
            PyObject *type2, *value2, *traceback2;
            PyErr_Fetch(&type2, &value2, &traceback2);
            PyErr_NormalizeException(&type2, &value2, &traceback2);
            PyException_SetContext(value2, value);
            PyErr_Restore(type2, value2, traceback2);
#else
            PyErr_SetFromErrno(PyExc_OSError);
#endif
        }
    }
    Py_XDECREF(types);
    return -1;
}

#ifdef IS_PY3

/**
 * Free the module's resources.
 *
 * Closes the socket.
 */
static void
arpreq_free(void *m)
{
    struct arpreq_state *st = GETSTATE(m);
    if (st->socket >= 0) {
        close(st->socket);
        st->socket = -1;
    }
}


/**
 * Traverse the references this module holds during GC
 */
static int
arpreq_traverse(PyObject *m, visitproc visit, void *arg)
{
    Py_VISIT(GETSTATE(m)->ipaddress_types);
    return 0;
}


/**
 * Clear the module.
 */
static int
arpreq_clear(PyObject *m)
{
    Py_CLEAR(GETSTATE(m)->ipaddress_types);
    return 0;
}

#ifdef HAVE_PEP489
static PyModuleDef_Slot arpreq_slots[] = {
    {Py_mod_exec, arpreq_exec},
    {0, NULL},
};
#endif

static struct PyModuleDef moduledef = {
        PyModuleDef_HEAD_INIT,
        "arpreq",
        arpreq_doc,
        sizeof(struct arpreq_state),
        arpreq_methods,
#ifdef HAVE_PEP489
        arpreq_slots,
#else
        NULL,
#endif
        arpreq_traverse,
        arpreq_clear,
        arpreq_free
};

#  define MOD_INIT(name) PyMODINIT_FUNC PyInit_##name(void)
#  define MOD_SUCCESS(module) return module
#  define MOD_ERROR(module) { Py_XDECREF(module); return NULL; }
#else /* IS_PY3 */
#  define MOD_INIT(name) PyMODINIT_FUNC init##name(void)
#  define MOD_SUCCESS(module) return
#  define MOD_ERROR(module) return
#endif

/**
 * Initialize the arpreq module
 */
MOD_INIT(arpreq)
{
#ifdef HAVE_PEP489
    return PyModuleDef_Init(&moduledef);
#else
#  ifdef IS_PY3
    PyObject *module = PyModule_Create(&moduledef);
#  else
    PyObject *module = Py_InitModule3("arpreq", arpreq_methods, arpreq_doc);
#  endif
    if (module == NULL) {
        goto fail;
    }
    if (arpreq_exec(module) == -1) {
        goto fail;
    }
    MOD_SUCCESS(module);
fail:
    MOD_ERROR(module);
#endif /* HAVE_PEP489 */
}
